lib/internal/modules/esm/loader.js

```diff
@@ -533,362 +277,74 @@ class ESMLoader {
-        if (replacedImportMetaInitializer) {
-          this.#importMetaInitializer = next;
-        }
-      }
-    }
-  }
-
-  importMetaInitialize(meta, context) {
-    this.#importMetaInitializer(meta, context);
-  }
-
   /**
    * Resolve the location of the module.
    *
-   * Internally, this behaves like a backwards iterator, wherein the stack of
-   * hooks starts at the top and each call to `nextResolve()` moves down 1 step
-   * until it reaches the bottom or short-circuits.
-   *
    * @param {string} originalSpecifier The specified URL path of the module to
    *                                   be resolved.
    * @param {string} [parentURL] The URL path of the module's parent.
    * @param {ImportAssertions} [importAssertions] Assertions from the import
    *                                              statement or expression.
-   * @returns {{ format: string, url: URL['href'] }}
+   * @returns {Promise<{ format: string, url: URL['href'] }>}
    */
   async resolve(
     originalSpecifier,
     parentURL,
     importAssertions = ObjectCreate(null),
   ) {
-    const isMain = parentURL === undefined;
-
-    if (
-      !isMain &&
-      typeof parentURL !== 'string' &&
-      !isURLInstance(parentURL)
-    ) {
-      throw new ERR_INVALID_ARG_TYPE(
-        'parentURL',
-        ['string', 'URL'],
-        parentURL,
-      );
+    if (this.#hooks) {
+      return this.#hooks.resolve(originalSpecifier, parentURL, importAssertions);
+    }
+    if (!this.#defaultResolve || !this.#defaultConditions) {
+      const { defaultResolve, DEFAULT_CONDITIONS } = require('internal/modules/esm/resolve');
+      this.#defaultResolve = defaultResolve;
+      this.#defaultConditions = DEFAULT_CONDITIONS;
     }
-    const chain = this.#hooks.resolve;
     const context = {
-      conditions: DEFAULT_CONDITIONS,
+      __proto__: null,
+      conditions: this.#defaultConditions,
       importAssertions,
       parentURL,
     };
-    const meta = {
-      chainFinished: null,
-      context,
-      hookErrIdentifier: '',
-      hookIndex: chain.length - 1,
-      hookName: 'resolve',
-      shortCircuited: false,
-    };
+    return this.#defaultResolve(originalSpecifier, context);
 
-    const validateArgs = (hookErrIdentifier, suppliedSpecifier, ctx) => {
-      validateString(
-        suppliedSpecifier,
-        `${hookErrIdentifier} specifier`,
-      ); // non-strings can be coerced to a url string
+  }
 
-      if (ctx) validateObject(ctx, `${hookErrIdentifier} context`);
-    };
-    const validateOutput = (hookErrIdentifier, output) => {
-      if (typeof output !== 'object' || output === null) { // [2]
-        throw new ERR_INVALID_RETURN_VALUE(
-          'an object',
-          hookErrIdentifier,
-          output,
-        );
+  /**
+   * Provide source that is understood by one of Node's translators.
+   *
+   * @param {URL['href']} url The URL/path of the module to be loaded
+   * @param {object} context Metadata about the module
```

Comment:
We should mark that the argument is optional.
```suggestion
   * @param {object} [context] Metadata about the module
```